File: kotlin/src/test/kotlin/com/bhengubv/circleai/personalfinance/PersonalFinanceTest.kt
// PersonalFinanceTest.kt
//
// Verifies the CircleAI.Personal.Finance port against the C# reference:
//   - upsert/getAccount; record (unknown account throws) mutates balance
//   - listForMonth filters by account + UTC year/month
//   - setBudget is case-insensitive on category; budgets ordered by category ASC
//   - summarise: byCategory signed sums, totalIn = Σ positive, totalOut = −Σ negative
//   - domain-context constants; adapter enrichment + a domain helper

package com.bhengubv.circleai.personalfinance

import com.bhengubv.circleai.companion.support.FakeCompanionSession
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Test
import java.math.BigDecimal
import java.time.Instant
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNull
import kotlin.test.assertTrue

class PersonalFinanceTest {

    private fun money(s: String) = BigDecimal(s)
    private fun assertMoney(expected: String, actual: BigDecimal) =
        assertTrue(BigDecimal(expected).compareTo(actual) == 0, "expected $expected but was $actual")

    private fun tx(id: String, amt: String, cat: String, at: Instant) =
        FinanceTransaction(id, "a1", money(amt), cat, null, at)

    private val jul = Instant.parse("2026-07-15T00:00:00Z")

    @Test
    fun `record requires known account and mutates balance`() {
        val b = InMemoryPersonalFinanceBoard()
        b.upsert(Account("a1", "Cheque", money("100"), "ZAR"))
        assertEquals("Cheque", b.getAccount("a1")!!.name)
        assertNull(b.getAccount("x"))

        b.record(tx("t1", "-40", "food", jul))
        assertMoney("60", b.getAccount("a1")!!.balance)
        assertFailsWith<IllegalStateException> { b.record(tx("bad", "1", "x", jul).copy(accountId = "ghost")) }
    }

    @Test
    fun `budgets case-insensitive and ordered`() {
        val b = InMemoryPersonalFinanceBoard()
        b.setBudget(BudgetLine("Food", money("2000")))
        b.setBudget(BudgetLine("food", money("2500"))) // overwrites (case-insensitive key)
        b.setBudget(BudgetLine("Airtime", money("300")))
        val budgets = b.budgets
        assertEquals(listOf("Airtime", "food"), budgets.map { it.category })
        assertMoney("2500", budgets.first { it.category.equals("food", ignoreCase = true) }.monthlyLimit)
    }

    @Test
    fun `summarise splits in and out and groups by category`() {
        val b = InMemoryPersonalFinanceBoard()
        b.upsert(Account("a1", "Cheque", money("0"), "ZAR"))
        b.record(tx("t1", "5000", "salary", jul))
        b.record(tx("t2", "-1200", "rent", jul))
        b.record(tx("t3", "-300", "food", jul))
        b.record(tx("t4", "-200", "food", jul))
        // a transaction in a different month must be excluded
        b.record(tx("t5", "-999", "food", Instant.parse("2026-06-15T00:00:00Z")))

        val s = b.summarise("a1", 2026, 7)
        assertEquals(2026, s.year)
        assertEquals(7, s.month)
        assertMoney("5000", s.totalIn)
        assertMoney("1700", s.totalOut) // 1200 + 300 + 200
        assertMoney("-500", s.byCategory.getValue("food")) // -300 + -200
        assertMoney("5000", s.byCategory.getValue("salary"))
        assertEquals(2, b.listForMonth("a1", 2026, 7).count { it.category == "food" })
    }

    @Test
    fun `domain context and adapter`() = runTest {
        assertTrue(PersonalFinanceDomainContext.SYSTEM_PROMPT_SNIPPET.startsWith("[DOMAIN: Personal.Finance]"))
        assertTrue("Not_Financial_Advice" in PersonalFinanceDomainContext.complianceFlags)

        val fake = FakeCompanionSession()
        val a = PersonalFinanceCompanionAdapter(fake)
        a.sendAsync("hi")
        assertTrue(fake.lastMessage!!.startsWith("[DOMAIN: Personal.Finance]"))
        a.createDebtPlanAsync("card R5000 @ 22%")
        assertTrue(fake.lastMessage!!.contains("avalanche method"))
    }
}

File: rust/tests/personal_finance_test.rs
//! personal_finance_test.rs
//!
//! Ports the behaviour of `CircleAI.Personal.Finance`: accounts + balance
//! application on record, unknown-account guard, per-month listing,
//! case-insensitive budgets ordered by category, the month summary (in/out +
//! by-category), the static domain descriptor, and the domain adapter.

use std::cell::RefCell;

use chrono::{TimeZone, Utc};
use circle_ai::companion::types::{
    CompanionContext, CompanionTurn, ICompanionSession, InterfaceKind,
};
use circle_ai::personal_finance::{
    Account, BudgetLine, FinanceTransaction, IPersonalFinanceBoard, InMemoryPersonalFinanceBoard,
    PersonalFinanceCompanionAdapter, PersonalFinanceDomainContext,
};

fn at(y: i32, m: u32, day: u32) -> chrono::DateTime<Utc> {
    Utc.with_ymd_and_hms(y, m, day, 9, 0, 0).unwrap()
}

#[test]
fn record_applies_balance() {
    let board = InMemoryPersonalFinanceBoard::new();
    board.upsert(Account::new("a1", "Cheque", 100.0, "ZAR"));
    board.record(FinanceTransaction::new("t1", "a1", -30.0, "Food", None, at(2026, 5, 1)));
    board.record(FinanceTransaction::new("t2", "a1", 50.0, "Salary", None, at(2026, 5, 2)));
    assert!((board.get_account("a1").unwrap().balance - 120.0).abs() < 1e-9);
}

#[test]
#[should_panic(expected = "Unknown account")]
fn record_unknown_account_panics() {
    let board = InMemoryPersonalFinanceBoard::new();
    board.record(FinanceTransaction::new("t1", "nope", 1.0, "X", None, at(2026, 5, 1)));
}

#[test]
fn list_for_month_filters_by_account_and_month() {
    let board = InMemoryPersonalFinanceBoard::new();
    board.upsert(Account::new("a1", "Cheque", 0.0, "ZAR"));
    board.record(FinanceTransaction::new("may1", "a1", 10.0, "X", None, at(2026, 5, 3)));
    board.record(FinanceTransaction::new("may2", "a1", 20.0, "Y", None, at(2026, 5, 20)));
    board.record(FinanceTransaction::new("jun", "a1", 99.0, "Z", None, at(2026, 6, 1)));

    let ids: Vec<String> = board
        .list_for_month("a1", 2026, 5)
        .into_iter()
        .map(|t| t.tx_id)
        .collect();
    assert_eq!(ids.len(), 2);
    assert!(ids.contains(&"may1".to_string()) && ids.contains(&"may2".to_string()));
}

#[test]
fn budgets_case_insensitive_key_ordered_by_category() {
    let board = InMemoryPersonalFinanceBoard::new();
    board.set_budget(BudgetLine::new("Food", 2000.0));
    board.set_budget(BudgetLine::new("food", 2500.0)); // overwrites (case-insensitive key)
    board.set_budget(BudgetLine::new("Airtime", 300.0));

    let budgets = board.budgets();
    let cats: Vec<String> = budgets.iter().map(|b| b.category.clone()).collect();
    assert_eq!(cats, vec!["Airtime", "food"]);
    // The overwrite kept the latest value + latest original casing.
    let food = budgets.iter().find(|b| b.category.eq_ignore_ascii_case("food")).unwrap();
    assert_eq!(food.monthly_limit, 2500.0);
}

#[test]
fn summarise_splits_in_out_and_by_category() {
    let board = InMemoryPersonalFinanceBoard::new();
    board.upsert(Account::new("a1", "Cheque", 0.0, "ZAR"));
    board.record(FinanceTransaction::new("t1", "a1", 100.0, "Salary", None, at(2026, 4, 1)));
    board.record(FinanceTransaction::new("t2", "a1", -30.0, "Food", None, at(2026, 4, 2)));
    board.record(FinanceTransaction::new("t3", "a1", -20.0, "Food", None, at(2026, 4, 3)));

    let s = board.summarise("a1", 2026, 4);
    assert_eq!(s.year, 2026);
    assert_eq!(s.month, 4);
    assert!((s.total_in - 100.0).abs() < 1e-9);
    assert!((s.total_out - 50.0).abs() < 1e-9);
    assert!((s.by_category["Food"] + 50.0).abs() < 1e-9);
    assert!((s.by_category["Salary"] - 100.0).abs() < 1e-9);
}

#[test]
fn domain_context_snippet_and_flags() {
    assert!(PersonalFinanceDomainContext::SYSTEM_PROMPT_SNIPPET
        .starts_with("[DOMAIN: Personal.Finance]"));
    assert!(PersonalFinanceDomainContext::SYSTEM_PROMPT_SNIPPET.contains("not advice"));
    assert_eq!(
        PersonalFinanceDomainContext::compliance_flags(),
        vec!["FAIS_Act_37_2002", "NCA", "POPIA", "Not_Financial_Advice"]
    );
    assert_eq!(
        PersonalFinanceDomainContext::suggested_tools(),
        vec!["budget_tracker", "spreadsheet", "calculator", "web_search"]
    );
}

// ── Adapter fixture ──────────────────────────────────────────────────────────

#[derive(Debug)]
struct FakeError(String);
impl std::fmt::Display for FakeError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}
impl std::error::Error for FakeError {}

struct RecordingSession {
    context: CompanionContext,
    history: Vec<CompanionTurn>,
    last_send: RefCell<Option<String>>,
    last_agent: RefCell<Option<String>>,
}

impl RecordingSession {
    fn new() -> Self {
        Self {
            context: CompanionContext::new(
                "id-1",
                "User",
                None,
                InterfaceKind::Mobile,
                "",
                "",
                Vec::new(),
                Vec::new(),
            ),
            history: vec![CompanionTurn::user("hi")],
            last_send: RefCell::new(None),
            last_agent: RefCell::new(None),
        }
    }
}

impl ICompanionSession for RecordingSession {
    type Error = FakeError;
    fn session_id(&self) -> &str {
        "sess-pfin"
    }
    fn identity_id(&self) -> &str {
        "id-1"
    }
    fn interface(&self) -> InterfaceKind {
        InterfaceKind::Mobile
    }
    fn send(&mut self, message: &str) -> Result<String, Self::Error> {
        *self.last_send.borrow_mut() = Some(message.to_string());
        Ok(format!("echo:{message}"))
    }
    fn stream(
        &mut self,
        message: &str,
    ) -> Result<Box<dyn Iterator<Item = Result<String, Self::Error>>>, Self::Error> {
        Ok(Box::new(std::iter::once(Ok(format!("chunk:{message}")))))
    }
    fn agent(&mut self, instruction: &str) -> Result<String, Self::Error> {
        *self.last_agent.borrow_mut() = Some(instruction.to_string());
        Ok(format!("agent:{instruction}"))
    }
    fn get_context(&self) -> &CompanionContext {
        &self.context
    }
    fn refresh_context(&mut self) -> Result<(), Self::Error> {
        Ok(())
    }
    fn history(&self) -> &[CompanionTurn] {
        &self.history
    }
    fn signal_feedback(&mut self, _p: bool, _n: Option<&str>) -> Result<(), Self::Error> {
        Ok(())
    }
}

#[test]
fn adapter_prefixes_domain_snippet_on_send() {
    let mut adapter = PersonalFinanceCompanionAdapter::new(RecordingSession::new());
    adapter.send("help").unwrap();
    let seen = adapter.inner().last_send.borrow().clone().unwrap();
    assert_eq!(
        seen,
        format!("{}\n\nhelp", PersonalFinanceDomainContext::SYSTEM_PROMPT_SNIPPET)
    );
}

#[test]
fn adapter_metadata_passthrough() {
    let adapter = PersonalFinanceCompanionAdapter::new(RecordingSession::new());
    assert_eq!(adapter.session_id(), "sess-pfin");
    assert_eq!(adapter.history().len(), 1);
}

#[test]
fn domain_helpers_use_raw_instructions() {
    let mut adapter = PersonalFinanceCompanionAdapter::new(RecordingSession::new());

    adapter.build_budget("R30k", "R25k").unwrap();
    let seen = adapter.inner().last_agent.borrow().clone().unwrap();
    assert!(!seen.starts_with(PersonalFinanceDomainContext::SYSTEM_PROMPT_SNIPPET));
    assert!(seen.contains("50/30/20 rule"));

    // Money rendered via the plain slot (integer -> no decimal point).
    adapter.design_savings_goal("Emergency fund", 30000.0, 12).unwrap();
    assert!(adapter
        .inner()
        .last_agent
        .borrow()
        .as_deref()
        .unwrap()
        .contains("save 30000 for 'Emergency fund' in 12 months"));
}
